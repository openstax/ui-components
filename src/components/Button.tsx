import { colors } from "../theme";
import styled from "styled-components";

const StyledButton = styled.button`
  background-color: ${colors.button.background};
  color: ${colors.palette.white};
  font-size: 1.6rem;
  font-weight: 700;
  line-height: 2rem;
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  height: 4rem;
  padding: 0 3rem;
  border: 0;
  border-radius: 5px;
  box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.2);

  &:not([disabled]) {
    cursor: pointer;
    &:hover {
      background: ${colors.button.backgroundHover}
    }
    &:active {
      background: ${colors.button.backgroundActive}
    }
  }
  &:disabled {
    opacity: 0.4;
  }
`;

interface ButtonProps extends React.ComponentPropsWithoutRef<'button'> {
  isWaiting?: never;
  waitingText?: never;
}
interface WaitingButtonProps extends React.ComponentPropsWithoutRef<'button'> {
  isWaiting: boolean;
  waitingText: string;
}

const Button = (props: ButtonProps | WaitingButtonProps) => {
  const { disabled, isWaiting, waitingText, children, ...otherProps } = props;

  return (
    <StyledButton {...otherProps} disabled={isWaiting || disabled}>{(isWaiting && waitingText) || children}</StyledButton>
  );
}

export default Button;
