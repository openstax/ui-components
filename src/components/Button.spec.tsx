import Button from './Button';
import renderer from 'react-test-renderer';

describe('Button', () => {
  it('matches snapshot', () => {
    const tree = renderer.create(
      <Button>Click Me</Button>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('isWaiting state matches snapshot', () => {
    const tree = renderer.create(
      <Button isWaiting={true} waitingText="Submitting...">Click Me</Button>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
