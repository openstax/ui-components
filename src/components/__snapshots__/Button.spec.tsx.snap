// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button isWaiting state matches snapshot 1`] = `
<button
  className="sc-bczRLJ fLTvYy"
  disabled={true}
>
  Submitting...
</button>
`;

exports[`Button matches snapshot 1`] = `
<button
  className="sc-bczRLJ fLTvYy"
>
  Click Me
</button>
`;
